clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,411 @@
+)
+
+const (
+	serviceCreationLatencyName           = "ServiceCreationLatency"
+	serviceCreationLatencyWorkers        = 10
+	defaultServiceCreationLatencyTimeout = 1 * time.Minute
+	defaultWaitTimeout                   = 10 * time.Minute
+	defaultCheckInterval                 = 10 * time.Second
+	pingBackoff                          = 1 * time.Second
+	pingChecks                           = 10
+)
+
+func init() {
+	measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement)
+}
+
+func createServiceCreationLatencyMeasurement() measurement.Measurement {
+	return &serviceCreationLatencyMeasurement{
+		queue:         workerqueue.NewWorkerQueue(serviceCreationLatencyWorkers),
+		creationTimes: make(map[string]*serviceCreationTimes),
+		pingCheckers:  make(map[string]*pingChecker),
+	}
+}
+
+type serviceCreationLatencyMeasurement struct {
+	namespace       string
+	labelSelector   string
+	fieldSelector   string
+	waitTimeout     time.Duration
+	stopCh          chan struct{}
+	isRunning       bool
+	selectorsString string
+	queue           workerqueue.Interface
+	client          clientset.Interface
+
+	lock          sync.Mutex
+	creationTimes map[string]*serviceCreationTimes
+	pingCheckers  map[string]*pingChecker
+}
+
+// Execute executes service startup latency measurement actions.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all-namespace scope is assumed.
+// "start" action starts observation of the services.
+// "waitForReady" waits until all services are reachable.
+// "gather" returns service created latency summary.
+func (s *serviceCreationLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		s.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return nil, err
+		}
+		s.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		s.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultServiceCreationLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.start()
+	case "waitForReady":
+		return nil, s.waitForReady()
+	case "gather":
+		return s.gather()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (s *serviceCreationLatencyMeasurement) Dispose() {
+	if s.isRunning {
+		s.isRunning = false
+		close(s.stopCh)
+	}
+	s.queue.Stop()
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for _, pc := range s.pingCheckers {
```

Comment:
This is also a common pattern.
I propose that we do the following:
- introduce Checker interface, that will only have a Stop() method
- on top of that we introduce type CheckersMap, that will have methods: NewCheckersMap and StopAllCheckers()

I would like to unify it further later, but what I mentioned above is simple and we can do that very quickly and will already introduce a common pattern for the future.

Reply:
this is still valid comment